src/module/custom.rs

```diff
@@ -0,0 +1,279 @@
+        I: Into<CustomSectionId>,
+    {
+        let id = id.into().0;
+        self.arena.delete(id);
+    }
+
+    /// Take a raw, unparsed custom section out of this module.
+    pub fn take_raw(&mut self, name: &str) -> Option<RawCustomSection> {
+        let id = self
+            .arena
+            .iter()
+            .filter(|(_id, s)| {
+                if let Some(s) = s {
+                    s.as_any().is::<RawCustomSection>() && s.name() == name
+                } else {
+                    false
+                }
+            })
+            .map(|(id, _)| id)
+            .next()?;
+        let section = self.arena[id].take().unwrap();
+        self.arena.delete(id);
+        let raw = section.into_any().downcast::<RawCustomSection>().unwrap();
+        Some(*raw)
+    }
+
+    /// Get a shared reference to a custom section that is in this
+    /// `ModuleCustomSections`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the custom section associated with the given `id` has been
+    /// deleted or taken.
+    pub fn get<T>(&self, id: TypedCustomSectionId<T>) -> &T
+    where
+        T: CustomSection,
+    {
+        self.try_get(id.into()).unwrap()
+    }
+
+    /// Try and get a shared reference to a custom section that is in this
+    /// `ModuleCustomSections`.
+    ///
+    /// Returns `None` if the section associated with the given `id` has been
+    /// taken or deleted, or if it is present but not an instance of `T`.
+    pub fn try_get<T>(&self, id: CustomSectionId) -> Option<&T>
+    where
+        T: CustomSection,
+    {
+        self.arena
+            .get(id.0)
+            .and_then(|s| s.as_ref().unwrap().as_any().downcast_ref::<T>())
+    }
+
+    /// Get an exclusive reference to a custom section that is in this
+    /// `ModuleCustomSections`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the custom section associated with the given `id` has been
+    /// deleted or taken.
+    pub fn get_mut<T>(&mut self, id: TypedCustomSectionId<T>) -> &mut T
+    where
+        T: CustomSection,
+    {
+        self.try_get_mut(id.into()).unwrap()
+    }
+
+    /// Try and get an exclusive reference to a custom section that is in this
+    /// `ModuleCustomSections`.
+    ///
+    /// Returns `None` if the section associated with the given `id` has been
+    /// taken or deleted, or if it is present but not an instance of `T`.
+    pub fn try_get_mut<T>(&mut self, id: CustomSectionId) -> Option<&mut T>
+    where
+        T: CustomSection,
+    {
+        self.arena
+            .get_mut(id.0)
+            .and_then(|s| s.as_mut().unwrap().as_any_mut().downcast_mut::<T>())
+    }
+
+    /// Get an untyped, shared reference to a custom section that is in this
+    /// `ModuleCustomSections`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the custom section associated with the given `id` has been
+    /// deleted or taken.
+    pub fn get_untyped(&self, id: CustomSectionId) -> &dyn CustomSection {
```

Comment:
So this is definitely a nitpick and not at all necessary for us to handle now. But I feel like this wants something like:

```rust
pub fn get<T>(&self, id: T) -> Option<&T::CustomSection>;
```

where you can `get(CustomSectionId)` to get `Option<&dyn CustomSection>` or you can `get(TypedSectionId<T>) -> Option<&T>`.

Basically just moving the typed/untyped distinction to the ID, and otherwise only having APIs that take ids and not some for typed ids and others for untyped ids.

Reply:
Nice idea.